File: Carla_Sumo_Code/TrustCal.py
import numpy as np
from vision import vision
from Behavior_Specs import dynamic_test, continuity, constraint_test


def calculate_trust(ego, simindex, que, table, order, k, trust_threshold, order_k):
    # Constants
    b = 1  # Set your value for 'b' here
    r = 1  # Set your value for 'r' here
    c = 5
    lambda_val = 0.95

    # Check conditions
    if sum(np.isnan(ego['scores'])) == 4 or ego['scores'][0] == 1:
        # Calculate scores
        score0 = continuity(ego)
        if score0 > 0:
            score1 = dynamic_test(ego)
            score2 = constraint_test(simindex, que, table, order, k, ego, trust_threshold)
            score3 = vision(que, ego, order_k)

            # Calculate trust
            scores = np.array([score0, score1, score2, score3])
            w = np.array([[0.6, 0.6, 0.6, 0.6], [1000, 10, 250, 0.25]])
            ego['reward'] = lambda_val * ego['reward'] + np.dot(w[0], scores)
            ego['regret'] = lambda_val * ego['regret'] + np.dot(w[1], 1 - scores)

            trust = ego['reward'] / (ego['reward'] + ego['regret'] + c)
        else:
            scores = np.array([0, 0, 0, 0])
            trust = 0
    else:
        scores = np.array([0, 0, 0, 0])
        trust = 0
        return scores, trust

    return scores, trust

File: Matlab code/TrustCal.m
function [ego,score,trust] = TrustCal(simindex,que,table,order,ego,order_k,k,CAV_e,trust_thereshold)
%UNTITLED3 Summary of this function goes here
%   Detailed explanation goes here

lambda =0.95;
if sum(isnan(ego.scores)) == 4 || (ego.scores(1) == 1)
    score0 = Continuity(ego);
    if score0 > 0
        score1 = DynamicTest(ego);
        score2 = ConstraintTest(simindex,que,table,order,k,ego,trust_thereshold);
        

        score3 = VisionTest(que,ego,order_k);
        score = [score0;score1;score2;score3];
        c = 5;
%         w = [0.6 0.6 0.6 0.6;1000 50 1000 0.25];
        w = [0.6 0.6 0.6 0.6;1000 10 250 0.25];
        ego.reward = lambda*ego.reward + w(1,:)*score;
        ego.regret = lambda*ego.regret + w(2,:)*(1-score);

        %trust = min(ego.trust(1) + b/(b+r+c),1);
        trust = ego.reward/(ego.reward+ego.regret+c);
    else
        score = [0,0,0,0];
        trust = 0;            
    end
else
    score = [0,0,0,0];
    trust = 0;    
    return
    
    
end


end

